OSDOCS-20040: Adds PlanPolicy to RHCL (#112905)
openshift/openshift-docs: modules/proc-deploying-planpolicy-rate-limiting.adoc

// Module included in the following assemblies:
//

> // *deployment/rhcl-PlanPolicy.adoc

Reviewer: ```suggestion
// *deployment/rhcl-planpolicy.adoc
```

:_mod-docs-content-type: PROCEDURE
[id="proc-deploying-planpolicy-rate-limiting_{context}"]

> = Deploying plan-based rate limiting with PlanPolicy

Reviewer: ```suggestion
= Deploying plan-based rate limiting with a PlanPolicy
```

[role="_abstract"]

> You can deploy a PlanPolicy resource to enforce tiered rate limits on an API based on authenticated user plan by using {prodname} on {product-title}.

Reviewer: ```suggestion
After you create your `AuthPolicy` and `RateLimitPolicy` custom resources (CRs), you can create a `PlanPolicy` object to enforce tiered rate limits on specific authenticated users.
```

> == Prerequisites

Reviewer: ```suggestion
.Prerequisites
```

> * {prodname} is installed in your OpenShift cluster. For more information, see the {prodname} installation documentation.

Reviewer: i think you can use the previous list, then just add that you created your auth and rate limit policies.

* The OpenShift CLI (`oc`) is installed and you are logged in to your cluster.
* A Gateway and an HTTPRoute are deployed for the application you want to protect.

> == Setting up environment variables

Reviewer: ```suggestion
```
none of this is needed; delete from here through line 246

          allNamespaces: true
        credentials:
          authorizationHeader:
            prefix: APIKEY
EOF
----

. Create API key secrets for each plan tier. Each secret carries a `secret.kuadrant.io/plan-id` annotation that identifies the plan:
+
[source,terminal,subs="+quotes"]
----
$ oc apply -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: gold-key
  namespace: ${KUADRANT_GATEWAY_NS}
  labels:
    authorino.kuadrant.io/managed-by: authorino
    app: toystore
  annotations:
    secret.kuadrant.io/plan-id: gold
stringData:
  api_key: _<gold-api-key-value>_
type: Opaque
---
apiVersion: v1
kind: Secret
metadata:
  name: silver-key
  namespace: ${KUADRANT_GATEWAY_NS}
  labels:
    authorino.kuadrant.io/managed-by: authorino
    app: toystore
  annotations:
    secret.kuadrant.io/plan-id: silver
stringData:
  api_key: _<silver-api-key-value>_
type: Opaque
---
apiVersion: v1
kind: Secret
metadata:
  name: bronze-key
  namespace: ${KUADRANT_GATEWAY_NS}
  labels:
    authorino.kuadrant.io/managed-by: authorino
    app: toystore
  annotations:
    secret.kuadrant.io/plan-id: bronze
stringData:
  api_key: _<bronze-api-key-value>_
type: Opaque
EOF
----

> == Deploying the PlanPolicy

Reviewer: ```suggestion
```

> . Deploy the PlanPolicy resource to enforce plan-based rate limits on the HTTPRoute:

Reviewer: ```suggestion
. Create a `PlanPolicy` object to enforce plan-based rate limits on the route:
```
let's ask if gRPC routes can be used with PlanPolicies (I would hope so since those are usually for mobile apps and devices AFIAK)

Author: It was confirmed that gRPC routes can be used with PlanPolicies.

+

> [source,terminal,subs="+quotes"]

Reviewer: ```suggestion
[source,yaml,subs="+quotes"]
```

----

> $ oc apply -f - <<EOF

Reviewer: ```suggestion
```

apiVersion: extensions.kuadrant.io/v1alpha1
kind: PlanPolicy
metadata:

>   name: toystore-plans

Reviewer: ```suggestion
  name: _<sample_app_planpolicy>_
```
oh, my bad i didn't say this earlier, but i usually just use whatever this value is + yaml when I do the apply
so here it would be _<sample_app_planpolicy.yaml>_

  namespace: ${KUADRANT_DEVELOPER_NS}
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute

>     name: toystore

Reviewer: ```suggestion
    name: _<sample_app>_
```
just replace all of these

  plans:
    - tier: gold
      predicate: |
        has(auth.identity) && auth.identity.metadata.annotations["secret.kuadrant.io/plan-id"] == "gold"
      limits:
        daily: 5
    - tier: silver
      predicate: |
        has(auth.identity) && auth.identity.metadata.annotations["secret.kuadrant.io/plan-id"] == "silver"
      limits:
        daily: 2
    - tier: bronze
      predicate: |
        has(auth.identity) && auth.identity.metadata.annotations["secret.kuadrant.io/plan-id"] == "bronze"
      limits:
        daily: 1

> EOF

Reviewer: ```suggestion
```

----
+

> Each `predicate` is a CEL expression. The predicates are evaluated in order, and the first matching plan is applied to the authenticated user.

Reviewer: ```suggestion
* Each `predicate` is a CEL expression. The predicates are evaluated in order, and the first matching plan is applied to the authenticated user.
```
add the whole path to predicate (i can't see through my comments)

+

> The `limits` field supports the following windows: `daily`, `weekly`, `monthly`, `yearly`, and `custom`. Custom limits take a `limit` value and a `window` duration string, for example:

Reviewer: ```suggestion
* The `limits` field supports the following windows: `daily`, `weekly`, `monthly`, `yearly`, and `custom`. Custom limits take a `limit` value and a `window` duration string, for example:
```

+
[source,yaml]
----
limits:
  daily: 100
  weekly: 500
  monthly: 2000
  custom:
    - limit: 10
      window: "1m"
----

> == Verification

Reviewer: move to separate module; all previous comments to verification steps and object formatting, etc, apply

> . Confirm that the AuthConfig was updated with plan data from the PlanPolicy:

Reviewer: ```suggestion
. Confirm that the `AuthConfig` CR was updated with plan data from the `PlanPolicy` object:
```
hmm, i don't recall if this is an auto-generated CR that we don't really see or what? let's ask
and yes, CR and object are the same thing, i just get bored; and i like to use `object` for the main idea

+
[source,terminal,subs="+quotes"]
----

> $ oc get authconfig -n kuadrant-system -o yaml

Reviewer: ```suggestion
$ oc get authconfig -n _<kuadrant_system>_ -o yaml
```
+
Replace `_<kuadrant_system>_` with the namespace you used for this {prodname} deployment.

> ----

Reviewer: (you can deploy many instances of RHCL across different namespaces; kind of like microshift in that way)

+

> The output should include plan information in the response section of the AuthConfig.

Reviewer: let's see if we can get a snippet of that output

> . Send a request without an API key and confirm that it is rejected:

Reviewer: ```suggestion
. Send a request without an API key and confirm that it is rejected by...:
```
you know...

+
[source,terminal,subs="+quotes"]
----
$ curl -H 'Host: _<your-api-hostname>_' http://${KUADRANT_GATEWAY_HOSTNAME}/toy -i
----
+
.Expected output
[source,terminal,subs="+quotes"]
----
HTTP/1.1 401 Unauthorized
----

. Send a request using the bronze-tier API key and confirm that it succeeds:
+
[source,terminal,subs="+quotes"]
----
$ curl -H 'Host: _<your-api-hostname>_' -H 'Authorization: APIKEY _<bronze-api-key-value>_' http://${KUADRANT_GATEWAY_HOSTNAME}/toy -i
----
+
.Expected output
[source,terminal,subs="+quotes"]
----
HTTP/1.1 200 OK
----

. Send a second request using the bronze-tier key and confirm that it is rate limited:
+
[source,terminal,subs="+quotes"]
----
$ curl -H 'Host: _<your-api-hostname>_' -H 'Authorization: APIKEY _<bronze-api-key-value>_' http://${KUADRANT_GATEWAY_HOSTNAME}/toy -i
----
+
.Expected output
[source,terminal,subs="+quotes"]
----
HTTP/1.1 429 Too Many Requests
----

. Confirm that the gold-tier key allows up to five requests per day before the limit is enforced.

> == Cleaning up

Reviewer: all of this to the end can be deleted; it assumes you were testing Kuadrant for something. not relevant to production use
